### ISSUE TITLE
Replace 'string.Template(..).substitute(..)' usage with 'str.format'

### ISSUE THREAD

**Issue by frewsxcv** [servo/servo#8857]
[`string.Template`](https://docs.python.org/2/library/string.html#template-strings) was introduced in Python 2.4. [`str.format`](https://docs.python.org/2/library/stdtypes.html#str.format) (a method on strings) was introduced in Python 2.6. Functionally, they are very similar, though the latter is less verbose.

e.g.

```python
string.Template("Hello ${arg}").substitute({"arg": "world!")
```

can be written as:

```python
"Hello {arg}".format(arg="world!")

# or

a = {"arg": "world!"}
"Hello {arg}".format(**a)
```

To find usages:

```
./mach grep string.Template
```

**Comment by Ms2ger** [comment]
So is this the recommended approach now?

**Comment by nox** [comment]
I just use `str % args` in CodegenRust, whenever I can. We should decide once and for all what to use.

**Comment by frewsxcv** [comment]
Unfortunately, there are many ways to do string formatting in Python (including a [new one coming in Python 3.6](https://www.python.org/dev/peps/pep-0498/)). I personally don't see any benefits for choosing `string.Template` over `str.format`, they're both pretty similar, one is just more verbose. The semantics of `%` formatting are quite different from `string.Template` and `str.format`, so replacing usages would be less trivial than the task in this issue. I personally don't have any issues with `%` and `str.format` coexisting.

**Comment by KiChjang** [comment]
In particular, using `%` formatting requires you to know the type of the arguments being substituted in order to work. Both `str.format` and `string.Template` are type-agnostic.

**Comment by ceph3us** [comment]
Since this was marked as a good first contribution, I've gone ahead and started work on this. Any objections?

**Comment by frewsxcv** [comment]
Go for it!

**Comment by ceph3us** [comment]
A few observations:
* `components/script/dom/bindings/codegen/GenerateCSS2PropertiesWebIDL.py` uses `string.Template` to fill out WebIDL templates for CSS2 properties in files. Obviously, the templating semantics are changed so these need to be modified. However, I can't find any of these in-tree and a few attempts at trying to dig them up by `grep`ing (both via `./mach grep` and `grep` directly) have been unsuccessful.

* `components/script/dom/bindings/codegen/CodegenRust.py` has a helper function called `fill()` that is implemented by means of `string.Template`. Should we try changing this or is it significantly abstract from the rest of the code that it should be left alone?

* If we leave the `fill()` function alone, should other instances of `string.Template` be left alone in that file for consistency?

**Comment by frewsxcv** [comment]
For your first point: https://github.com/servo/servo/issues/8998

For your second and third points, I would change everything to use `str.format` for consistency.. When I checked a week or so ago, it seemed like we did use any syntax that wasn't compatible between `string.Template` and `str.format`.

**Comment by Ms2ger** [comment]
Ignore `fill()`, but update the rest of `CodegenRust.py`, please.

**Comment by ceph3us** [comment]
This might not be desirable at all for `CodegenRust.py`; every generated Rust codeblock containing curly braces will need them escaped. The `{{` and `}}` everywhere for Rust codeblocks combined with the use of `{subst}` for string format parameters may not be a suitable choice for this usecase. I'll keep working on this for now in case everyone wants to proceed anyway.

**Comment by frewsxcv** [comment]
> This might not be desirable at all for CodegenRust.py; every generated Rust codeblock containing curly braces will need them escaped. The {{ and }} everywhere for Rust codeblocks combined with the use of {subst} for string format parameters may not be a suitable choice for this usecase. 

I don't follow. Can you share a concrete example of what you're talking about?

**Comment by ceph3us** [comment]
Take this particularly egregious example from `CodegenRust.py:5597` (as of d9bf1ba)

```python
string.Template("""\
impl CallbackContainer for ${type} {
    fn new(callback: *mut JSObject) -> Rc<${type}> {
        ${type}::new(callback)
    }
    fn callback(&self) -> *mut JSObject {
        self.parent.callback()
    }
}
impl ToJSValConvertible for ${type} {
    unsafe fn to_jsval(&self, cx: *mut JSContext, rval: MutableHandleValue) {
        self.callback().to_jsval(cx, rval);
    }
}\
""").substitute({"type": callback.identifier.name})
        CGGeneric.__init__(self, impl)
```
If you naively strip the $ from the substitutions and switch to `str.format`, you end up with exceptions thrown along the lines of `KeyError: '\n        {type}::new(callback)' because it interprets the codeblocks as substitutions.

The actual correct template then looks like this:
```python
impl = """\
impl CallbackContainer for {type} {{
    fn new(callback: *mut JSObject) -> Rc<{type}> {{
        {type}::new(callback)
    }}

    fn callback(&self) -> *mut JSObject {
        self.parent.callback()
    }}
}}

impl ToJSValConvertible for {type} {{
    unsafe fn to_jsval(&self, cx: *mut JSContext, rval: MutableHandleValue) {{
        self.callback().to_jsval(cx, rval);
    }}
}}\
"""
```

As you can see, that's not super fun to work with.

**Comment by KiChjang** [comment]
I'm not quite seeing why it's "not super fun to work with". Isn't that even a stronger reminder to the reader that we're working with Rust code when `{{` and `}}` are encountered?

**Comment by ceph3us** [comment]
YMMV. I found the two to not be so visually distinct, but I'm continuing to work on it anyway. Just making sure that the changes work for everyone.

**Comment by frewsxcv** [comment]
I sort of agree with @ceph3us that they're not fun to work with. I think the Python strings should be as close as possible to the outputted Rust code. When writing new Python code in this file, I'm fairly certain people are going to commonly run into braces issues frequently via `str.format`. I overlooked this difference with how braces work between `string.Template` and `str.format` when I first filed this issue.

That said, I'm now skeptical of the benefits of completing this issue. @ceph3us since you've already edited the files a bit, you can make a suggestive call whether you think the changes make the code cleaner. Otherwise, I'm sorry for misleading you down this path. Thanks for the effort though!

**Comment by ceph3us** [comment]
I think that the call of whether, from a purely Python code perspective, it's an improvement: less needless verbosity, more Pythonic. But the problem comes if someone who's knowledge is primarily in Rust wants to make a minor tweak to the generated DOM bindings, who doesn't know the particulars of `str.format`, accidently uses `{` instead of `{{` and they're left with an error message that gives them very little clue what happened. If that's not likely to be a common case, it would probably be fine to make the switch.

If you want to evaluate, I just pushed the changes to ceph3us/servo@a86b59f.

**Comment by frewsxcv** [comment]
I agree with your summary though: from a Python perspective, it's subjectively more idiomatic and slightly cleaner. But considering the implications of this change, it seems to be slightly more inconvenient to write Rust code generation, which I did not anticipate when I filed this issue. I fall somewhere between neutral and slightly against the changes, though someone with more opinions can override my opinion (though I think @Ms2ger and @nox would also be against the changes).

Since you seem to be familiar with Python (and now this file :stuck_out_tongue:) there's plenty of opportunities for cleaning up the Rust codegen
